index.ts

```diff
@@ -5,41 +5,62 @@ import fetchRetry from 'fetch-retry';
 import isString from 'lodash.isstring';
 import md5 from 'md5';
 import ms from 'ms';
-import fetch from 'node-fetch';
+import fetch, { Response } from 'node-fetch';
 import removeTrailingSlash from 'remove-trailing-slash';
 import { v4 as uuid } from 'uuid';
 
 const version = require('./package.json').version;
 
 const retryableFetch = fetchRetry(fetch as any) as unknown as typeof fetch;
 const setImmediate = global.setImmediate || process.nextTick.bind(process);
-const noop = () => {};
 
-enum AnalyticsState {
+export type AnalyticsMessage = AnalyticsIdentity & {
+  context?: { [key: string]: unknown };
+  traits?: { [key: string]: unknown };
+  integrations?: { [destination: string]: boolean };
+  timestamp?: Date;
+  [key: string]: unknown;
+};
+
+export type AnalyticsIdentity = { userId: string } | { anonymousId: string };
+
+export type AnalyticsMessageCallback = (error?: Error) => void;
+
+export type AnalyticsFlushCallback = (
+  error?: Error,
+  data?: { batch: AnalyticsPayload[]; sentAt: Date }
+) => void;
+
+type AnalyticsPayload = any;
+
+type AnalyticsEventType = 'identify' | 'track' | 'page' | 'screen' | 'group' | 'alias';
+
+const enum AnalyticsState {
   Idle = 'idle',
   Running = 'running',
 }
 
-type AnalyticsMessage = any;
-
-class Analytics {
+export default class Analytics {
   private readonly enable: boolean;
+
   private state = AnalyticsState.Idle;
-  private queue = [] as {
-    message: AnalyticsMessage;
-    callback: (error?: Error, data?: { batch: AnalyticsMessage[]; sentAt: Date }) => void;
+  private readonly queue = [] as {
+    message: AnalyticsPayload;
+    callback: AnalyticsMessageCallback;
   }[];
-  private writeKey: string;
-  private host: string;
-  private timeout: number | null;
-  private flushAt: number;
-  private flushInterval: number;
-  private maxInternalQueueSize: number;
+
+  private readonly writeKey: string;
+  private readonly host: string;
+  private readonly timeout: number | string | null;
+
+  private readonly flushAt: number;
+  private readonly flushInterval: number;
+  private readonly maxInternalQueueSize: number;
   private flushed: boolean = false;
-  private flushTimer: NodeJS.Timer | null = null;
   private timer: NodeJS.Timer | null = null;
+  private flushTimer: NodeJS.Timer | null = null;
```

Comment:
imo - prefer `private flushTimer?: NodeJS.Timer`

Reply:
The property exists on the instance so "?" technically works when `exactOptionalPropertyTypes` is disabled but doesn't follow the sprit of "?" since the field is never optional - it always exists. The point of using `null` though is to treat `undefined` as a programming error. It's easy to have the value be `undefined` on accident, harder for it to be null.

Reply:
Interesting. I've never considered `?` as optional when used on a class property. I've always looked at it as a `Nullable<T>`, maybe from my days programming in C#. I do like your PoV on `undefined` vs `null`, definitely makes sense. Thanks for the explanation, I think this typing is g2g, then.

Reply:
C#'s `?` is a little different since there are only nullable and non-nullable values.

With JS there's `null`, `undefined`, property not defined on an object (`Object.hasOwn(obj, 'x')`), property not defined on a prototype chain (`'x' in obj`). In TypeScript's case, `?` by default allows `undefined` or properties not defined on the prototype chain.

With `exactOptionalPropertyTypes`, `undefined` is disallowed and the property needs to be absent from the prototype chain.